File: src/keyboard_to_joy.cpp
#include "ros/ros.h"
#include "sensor_msgs/Joy.h"
#include "keyboard/Key.h"

#include "single_key.h"

class KeyboardToJoyNode
{

  SingleKey key_up;
  SingleKey key_down;
  SingleKey key_left;
  SingleKey key_right;
  ros::Publisher pub;
  ros::Subscriber sub_keyupp;
  ros::Subscriber sub_keydwn;
  ros::Timer timer;

public:

  KeyboardToJoyNode(int argc, char** argv);
  ~KeyboardToJoyNode();

  void readKeyUp(const keyboard::Key::ConstPtr& msg);
  void readKeyDown(const keyboard::Key::ConstPtr& msg);
  void updateOperatorSignal(const ros::TimerEvent& event);
  void spin();

};

KeyboardToJoyNode::KeyboardToJoyNode(int argc, char** argv)
{
  ros::init(argc, argv, "keyboard_to_joy");
  key_up = SingleKey(273);
  key_down = SingleKey(274);
  key_left = SingleKey(276);
  key_right = SingleKey(275);
  ros::NodeHandle nh;
  ros::NodeHandle nh_param("~");
  int hz;
  nh_param.param<int>("hz", hz, 100);
  sub_keyupp = nh.subscribe("keyboard/keyup", 1000, &KeyboardToJoyNode::readKeyUp, this);
  sub_keydwn = nh.subscribe("keyboard/keydown", 1000, &KeyboardToJoyNode::readKeyDown, this);
  pub = nh.advertise<sensor_msgs::Joy>("joy_out", 1000);
  timer = nh.createTimer(ros::Duration(1.0/static_cast<double>(hz)), &KeyboardToJoyNode::updateOperatorSignal, this);
}

KeyboardToJoyNode::~KeyboardToJoyNode()
{
  timer.stop();
  sub_keyupp.shutdown();
  sub_keydwn.shutdown();
}

void KeyboardToJoyNode::readKeyUp(const keyboard::Key::ConstPtr& msg)
{
  key_up.releasedKey(msg);
  key_down.releasedKey(msg);
  key_left.releasedKey(msg);
  key_right.releasedKey(msg);
}

void KeyboardToJoyNode::readKeyDown(const keyboard::Key::ConstPtr& msg)
{
  key_up.pressedKey(msg);
  key_down.pressedKey(msg);
  key_left.pressedKey(msg);
  key_right.pressedKey(msg);
}

void KeyboardToJoyNode::updateOperatorSignal(const ros::TimerEvent& event)
{
  double hori = key_left.getValue() - key_right.getValue();
  double vert = key_up.getValue() - key_down.getValue();
  sensor_msgs::Joy msg;
  msg.header.stamp = ros::Time::now();
  msg.axes.push_back(hori);
  msg.axes.push_back(vert);
  pub.publish(msg);
}

void KeyboardToJoyNode::spin()
{
  ros::spin();
}

int main(int argc, char** argv)
{
  KeyboardToJoyNode node = KeyboardToJoyNode(argc, argv);
  node.spin();
  return 0;
}

File: scripts/keyboard_to_joy.py
#!/usr/bin/env python3
# BSD 2-Clause License
# Copyright (c) 2022, Christopher E. Mower
# All rights reserved.
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
# 1. Redistributions of source code must retain the above copyright notice, this
#    list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright notice,
#    this list of conditions and the following disclaimer in the documentation
#    and/or other materials provided with the distribution.
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
import sys
import rospy
from sensor_msgs.msg import Joy
from keyboard.msg import Key


class Node:

    def __init__(self):

        # Initialize ROS
        rospy.init_node('keyboard_to_joy_node')

        # Setup publisher
        self.pub = rospy.Publisher('joy', Joy, queue_size=10)

        # Get parameters
        config = rospy.get_param('~config')
        self.axes_neg_code = [getattr(Key, "KEY_%s"%a[0]) for a in config['axes']]
        self.axes_pos_code = [getattr(Key, "KEY_%s"%a[1]) for a in config['axes']]
        self.buttons_code = [getattr(Key, "KEY_%s"%k) for k in config.get('buttons', [])]

        # Setup joy message
        self.msg = Joy(axes=[0.0]*len(config['axes']), buttons=[0]*len(self.buttons_code))

        # Start subscribers
        rospy.Subscriber('keyboard/keyup', Key, self.callback_keyup)
        rospy.Subscriber('keyboard/keydown', Key, self.callback_keydown)

        # Start timer
        dt = 1.0/float(rospy.get_param('~hz', 100))
        rospy.Timer(rospy.Duration(dt), self.main_loop)

        rospy.loginfo('initialized keyboard_to_joy node')

    def _update_buttons(self, msg, value):
        if msg.code in self.buttons_code:
            self.msg.buttons[self.buttons_code.index(msg.code)] = value
            return True
        return False

    def _update_axes(self, msg, axes_code, value):
        if msg.code in axes_code:
            self.msg.axes[axes_code.index(msg.code)] += value
            return True
        return False


    def callback_keyup(self, msg):
        if self._update_buttons(msg, 0):
            return
        if self._update_axes(msg, self.axes_neg_code, 1.0):
            return
        if self._update_axes(msg, self.axes_pos_code, -1.0):
            return

    def callback_keydown(self, msg):
        if self._update_buttons(msg, 1):
            return
        if self._update_axes(msg, self.axes_neg_code, -1.0):
            return
        if self._update_axes(msg, self.axes_pos_code, 1.0):
            return

    def main_loop(self, event):
        self.msg.header.stamp = rospy.Time.now()
        self.pub.publish(self.msg)

    def spin(self):
        rospy.spin()


def main():
    Node().spin()


if __name__ == '__main__':
    main()
